<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Data Frames</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">seandavi(s12): Courses and Tutorials</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="setup.html">
    <span class="fa fa-cogs"></span>
     
    setup
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="ion ion-easel"></span>
     
    Slides
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="motivation_for_R_slides.html">Motivation for using R</a>
    </li>
    <li>
      <a href="http://bit.ly/bioc_cshl_2019">Introduction to Bioconductor</a>
    </li>
    <li>
      <a href="https://doi.org/10.6084/m9.figshare.8397257.v1">Advanced Bioconductor Overview</a>
    </li>
    <li>
      <a href="MachineLearning.html">Machine Learning hands-on</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-question fa-lg"></span>
     
    Misc.
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="further_resources.html">Further resources</a>
    </li>
    <li>
      <a href="https://github.com/seandavi/ITR">Source code for this site</a>
    </li>
    <li>
      <a href="https://github.com/seandavi/ITR/archive/master.zip">Download materials</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Data Frames</h1>

</div>


<p>While R has many different data types, the one that is central to much of the power and popularity of R is the <code>data.frame</code>. A data.frame looks a bit like an R matrix in that it has two dimensions, rows and columns. However, data.frames are usually viewed as a set of columns representing variables and the rows representing the values of those variables. Importantly, a data.frame may contain <em>different</em> data types in each of its columns; matrices <strong>must</strong> contain only one data type. This distinction is important to remember, as there are <em>specific</em> approaches to working with R data.frames that may be different than those for working with matrices.</p>
<div id="learning-goals" class="section level2">
<h2>Learning goals</h2>
<ul>
<li>Understand how data.frames are different from matrices.</li>
<li>Know a few functions for examing the contents of a data.frame.</li>
<li>List approaches for subsetting data.frames.</li>
<li>Be able to load and save tabular data from and to disk.</li>
<li>Show how to create a data.frames from scratch.</li>
</ul>
</div>
<div id="learning-objectives" class="section level2">
<h2>Learning objectives</h2>
<ul>
<li>Load the yeast growth dataset into R using <code>read.csv</code>.</li>
<li>Examine the contents of the dataset.</li>
<li>Use subsetting to find genes that may be involved with nutrient metabolism and transport.</li>
<li>Summarize data measurements by categories.</li>
</ul>
</div>
<div id="dataset" class="section level2">
<h2>Dataset</h2>
<p>The data used here are borrowed directly from the <a href="http://bioconnector.org/index.html">fantastic Bioconnector tutorials</a> and are a cleaned up version of the data from <a href="http://www.ncbi.nlm.nih.gov/pubmed/17959824">Brauer et al. Coordination of Growth Rate, Cell Cycle, Stress Response, and Metabolic Activity in Yeast (2008) Mol Biol Cell 19:352-367</a>. These data are from a gene expression microarray, and in this paper the authors examine the relationship between growth rate and gene expression in yeast cultures limited by one of six different nutrients (glucose, leucine, ammonium, sulfate, phosphate, uracil). If you give yeast a rich media loaded with nutrients except restrict the supply of a single nutrient, you can control the growth rate to any rate you choose. By starving yeast of specific nutrients you can find genes that:</p>
<ol style="list-style-type: decimal">
<li>Raise or lower their expression in response to growth rate. Growth-rate dependent expression patterns can tell us a lot about cell cycle control, and how the cell responds to stress. The authors found that expression of &gt;25% of all yeast genes is linearly correlated with growth rate, independent of the limiting nutrient. They also found that the subset of negatively growth-correlated genes is enriched for peroxisomal functions, and positively correlated genes mainly encode ribosomal functions.</li>
<li>Respond differently when different nutrients are being limited. If you see particular genes that respond very differently when a nutrient is sharply restricted, these genes might be involved in the transport or metabolism of that specific nutrient.</li>
</ol>
<p>The dataset can be downloaded directly from:</p>
<ul>
<li><a href="https://raw.githubusercontent.com/bioconnector/workshops/master/data/brauer2007_tidy.csv">brauer2007_tidy.csv</a></li>
</ul>
<p>We are going to read this dataset into R and then use it as a playground for learning about data.frames.</p>
</div>
<div id="reading-in-data" class="section level2">
<h2>Reading in data</h2>
<p>R has many capabilities for reading in data. Many of the functions have names that help us to understand what data format is to be expected. In this case, the filename that we want to read ends in <code>.csv</code>, meaning comma-separated-values. The <code>read.csv()</code> function reads in <code>.csv</code> files. As usual, it is worth reading <code>help('read.csv')</code> to get a better sense of the possible bells-and-whistles.</p>
<p>The <code>read.csv()</code> function can read directly from a URL, so we do not need to download the file directly. This dataset is relatively large (about 16MB), so this may take a bit depending on your network connection speed.</p>
<pre class="r"><code>ydat &lt;- read.csv(&#39;https://raw.githubusercontent.com/bioconnector/workshops/master/data/brauer2007_tidy.csv&#39;)</code></pre>
<p>Our variable, <code>ydat</code>, now “contains” the downloaded and read data. We can check to see what data type read.csv gave us:</p>
<pre class="r"><code>class(ydat)</code></pre>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
</div>
<div id="inspecting-data.frames" class="section level2">
<h2>Inspecting data.frames</h2>
<p>Our <code>ydat</code> variable is a data.frame. As I mentioned, the dataset is fairly large, so we will not be able to look at it all at once on the screen. However, R gives us many tools to inspect a data.frame.</p>
<ul>
<li>Overviews of content
<ul>
<li><code>head()</code> to show first few rows</li>
<li><code>tail()</code> to show last few rows</li>
</ul></li>
<li>Size
<ul>
<li><code>dim()</code> for dimensions (rows, columns)</li>
<li><code>nrow()</code></li>
<li><code>ncol()</code></li>
<li><code>object.size()</code> for power users interested in the memory used to store an object</li>
</ul></li>
<li>Data and attribute summaries
<ul>
<li><code>colnames()</code> to get the names of the columns</li>
<li><code>rownames()</code> to get the “names” of the rows–may not be present</li>
<li><code>summary()</code> to get per-column summaries of the data in the data.frame.</li>
</ul></li>
</ul>
<pre class="r"><code>head(ydat)</code></pre>
<pre><code>##   symbol systematic_name nutrient rate expression
## 1   SFB2         YNL049C  Glucose 0.05      -0.24
## 2   &lt;NA&gt;         YNL095C  Glucose 0.05       0.28
## 3   QRI7         YDL104C  Glucose 0.05      -0.02
## 4   CFT2         YLR115W  Glucose 0.05      -0.33
## 5   SSO2         YMR183C  Glucose 0.05       0.05
## 6   PSP2         YML017W  Glucose 0.05      -0.69
##                             bp                            mf
## 1        ER to Golgi transport    molecular function unknown
## 2   biological process unknown    molecular function unknown
## 3 proteolysis and peptidolysis metalloendopeptidase activity
## 4      mRNA polyadenylylation*                   RNA binding
## 5              vesicle fusion*              t-SNARE activity
## 6   biological process unknown    molecular function unknown</code></pre>
<pre class="r"><code>tail(ydat)</code></pre>
<pre><code>##        symbol systematic_name nutrient rate expression
## 198425   DOA1         YKL213C   Uracil  0.3       0.14
## 198426   KRE1         YNL322C   Uracil  0.3       0.28
## 198427   MTL1         YGR023W   Uracil  0.3       0.27
## 198428   KRE9         YJL174W   Uracil  0.3       0.43
## 198429   UTH1         YKR042W   Uracil  0.3       0.19
## 198430   &lt;NA&gt;         YOL111C   Uracil  0.3       0.04
##                                                bp
## 198425    ubiquitin-dependent protein catabolism*
## 198426      cell wall organization and biogenesis
## 198427      cell wall organization and biogenesis
## 198428     cell wall organization and biogenesis*
## 198429 mitochondrion organization and biogenesis*
## 198430                 biological process unknown
##                                         mf
## 198425          molecular function unknown
## 198426 structural constituent of cell wall
## 198427          molecular function unknown
## 198428          molecular function unknown
## 198429          molecular function unknown
## 198430          molecular function unknown</code></pre>
<pre class="r"><code>dim(ydat)</code></pre>
<pre><code>## [1] 198430      7</code></pre>
<pre class="r"><code>nrow(ydat)</code></pre>
<pre><code>## [1] 198430</code></pre>
<pre class="r"><code>ncol(ydat)</code></pre>
<pre><code>## [1] 7</code></pre>
<pre class="r"><code>colnames(ydat)</code></pre>
<pre><code>## [1] &quot;symbol&quot;          &quot;systematic_name&quot; &quot;nutrient&quot;        &quot;rate&quot;           
## [5] &quot;expression&quot;      &quot;bp&quot;              &quot;mf&quot;</code></pre>
<pre class="r"><code>summary(ydat)</code></pre>
<pre><code>##      symbol       systematic_name       nutrient          rate       
##  AAC1   :    36   Q0017  :    36   Ammonia  :33141   Min.   :0.0500  
##  AAD10  :    36   Q0045  :    36   Glucose  :33138   1st Qu.:0.1000  
##  AAD14  :    36   Q0050  :    36   Leucine  :33178   Median :0.2000  
##  AAD16  :    36   Q0055  :    36   Phosphate:33068   Mean   :0.1752  
##  AAD4   :    36   Q0065  :    36   Sulfate  :32897   3rd Qu.:0.2500  
##  (Other):151000   Q0070  :    36   Uracil   :33008   Max.   :0.3000  
##  NA&#39;s   : 47250   (Other):198214                                     
##    expression                                             bp        
##  Min.   :-6.500000   biological process unknown            : 45139  
##  1st Qu.:-0.290000   protein biosynthesis                  :  6654  
##  Median : 0.000000   protein amino acid phosphorylation*   :  2792  
##  Mean   : 0.003367   protein biosynthesis*                 :  2624  
##  3rd Qu.: 0.290000   cell wall organization and biogenesis*:  2303  
##  Max.   : 6.640000   (Other)                               :131255  
##                      NA&#39;s                                  :  7663  
##                                   mf        
##  molecular function unknown        : 65280  
##  structural constituent of ribosome:  6723  
##  protein binding                   :  3981  
##  RNA binding                       :  2442  
##  protein binding*                  :  1903  
##  (Other)                           :110438  
##  NA&#39;s                              :  7663</code></pre>
<p>In RStudio, there is an additional function, <code>View()</code> (note the capital “V”) that opens the first 1000 rows (default) in the RStudio window, akin to a spreadsheet view.</p>
<pre class="r"><code>View(ydat)</code></pre>
</div>
<div id="accessing-variables-columns-and-subsetting" class="section level2">
<h2>Accessing variables (columns) and subsetting</h2>
<p>In R, data.frames can be subset similarly to other two-dimensional data structures. The <code>[</code> in R is used to denote subsetting of any kind. When working with two-dimensional data, we need two values inside the <code>[ ]</code> to specify the details. The specification is <code>[rows, columns]</code>. For example, to get the first three rows of <code>ydat</code>, use:</p>
<pre class="r"><code>ydat[1:3, ]</code></pre>
<pre><code>##   symbol systematic_name nutrient rate expression
## 1   SFB2         YNL049C  Glucose 0.05      -0.24
## 2   &lt;NA&gt;         YNL095C  Glucose 0.05       0.28
## 3   QRI7         YDL104C  Glucose 0.05      -0.02
##                             bp                            mf
## 1        ER to Golgi transport    molecular function unknown
## 2   biological process unknown    molecular function unknown
## 3 proteolysis and peptidolysis metalloendopeptidase activity</code></pre>
<p>Note how the second number, the columns, is blank. R takes that to mean “all the columns”. Similarly, we can combine rows and columns specification arbitrarily.</p>
<pre class="r"><code>ydat[1:3, 1:3]</code></pre>
<pre><code>##   symbol systematic_name nutrient
## 1   SFB2         YNL049C  Glucose
## 2   &lt;NA&gt;         YNL095C  Glucose
## 3   QRI7         YDL104C  Glucose</code></pre>
<p>Because selecting a single variable, or column, is such a common operation, there are two shortcuts for doing so <em>with data.frames</em>. The first, the <code>$</code> operator works like so:</p>
<pre class="r"><code># Look at the column names, just to refresh memory
colnames(ydat)</code></pre>
<pre><code>## [1] &quot;symbol&quot;          &quot;systematic_name&quot; &quot;nutrient&quot;        &quot;rate&quot;           
## [5] &quot;expression&quot;      &quot;bp&quot;              &quot;mf&quot;</code></pre>
<pre class="r"><code># Note that I am using &quot;head&quot; here to limit the output
head(ydat$symbol)</code></pre>
<pre><code>## [1] SFB2 &lt;NA&gt; QRI7 CFT2 SSO2 PSP2
## 4210 Levels: AAC1 AAC3 AAD10 AAD14 AAD16 AAD4 AAD6 AAH1 AAP1 AAP1&#39; ... ZWF1</code></pre>
<pre class="r"><code># What is the actual length of &quot;symbol&quot;?
length(ydat$symbol)</code></pre>
<pre><code>## [1] 198430</code></pre>
<p>The second is related to the fact that, in R, data.frames are also lists. We subset a list by using <code>[[]]</code> notation. To get the second column of <code>ydat</code>, we can use:</p>
<pre class="r"><code>head(ydat[[2]])</code></pre>
<pre><code>## [1] YNL049C YNL095C YDL104C YLR115W YMR183C YML017W
## 5536 Levels: Q0017 Q0045 Q0050 Q0055 Q0060 Q0065 Q0070 Q0075 Q0080 ... YPR204W</code></pre>
<p>Alternatively, we can use the column name:</p>
<pre class="r"><code>head(ydat[[&quot;systematic_name&quot;]])</code></pre>
<pre><code>## [1] YNL049C YNL095C YDL104C YLR115W YMR183C YML017W
## 5536 Levels: Q0017 Q0045 Q0050 Q0055 Q0060 Q0065 Q0070 Q0075 Q0080 ... YPR204W</code></pre>
<div id="some-data-exploration" class="section level3">
<h3>Some data exploration</h3>
<p>There are a couple of columns that include numeric values. Which columns are numeric?</p>
<pre class="r"><code>class(ydat$symbol)</code></pre>
<pre><code>## [1] &quot;factor&quot;</code></pre>
<pre class="r"><code>class(ydat$rate)</code></pre>
<pre><code>## [1] &quot;numeric&quot;</code></pre>
<pre class="r"><code>class(ydat$expression)</code></pre>
<pre><code>## [1] &quot;numeric&quot;</code></pre>
<p>Make histograms of: - the expression values - the rate values</p>
<p>What does the <code>table()</code> function do? Could you use that to look a the <code>rate</code> column given that that column appears to have repeated values?</p>
<p>What <code>rate</code> corresponds to the most nutrient-starved condition?</p>
</div>
<div id="more-advanced-indexing-and-subsetting" class="section level3">
<h3>More advanced indexing and subsetting</h3>
<p>We can use, for example, logical values (TRUE/FALSE) to subset data.frames.</p>
<pre class="r"><code>head(ydat[ydat$symbol == &#39;LEU1&#39;, ])</code></pre>
<pre><code>##      symbol systematic_name nutrient rate expression   bp   mf
## NA     &lt;NA&gt;            &lt;NA&gt;     &lt;NA&gt;   NA         NA &lt;NA&gt; &lt;NA&gt;
## NA.1   &lt;NA&gt;            &lt;NA&gt;     &lt;NA&gt;   NA         NA &lt;NA&gt; &lt;NA&gt;
## NA.2   &lt;NA&gt;            &lt;NA&gt;     &lt;NA&gt;   NA         NA &lt;NA&gt; &lt;NA&gt;
## NA.3   &lt;NA&gt;            &lt;NA&gt;     &lt;NA&gt;   NA         NA &lt;NA&gt; &lt;NA&gt;
## NA.4   &lt;NA&gt;            &lt;NA&gt;     &lt;NA&gt;   NA         NA &lt;NA&gt; &lt;NA&gt;
## NA.5   &lt;NA&gt;            &lt;NA&gt;     &lt;NA&gt;   NA         NA &lt;NA&gt; &lt;NA&gt;</code></pre>
<pre class="r"><code>tail(ydat[ydat$symbol == &#39;LEU1&#39;, ])</code></pre>
<pre><code>##          symbol systematic_name nutrient rate expression   bp   mf
## NA.47244   &lt;NA&gt;            &lt;NA&gt;     &lt;NA&gt;   NA         NA &lt;NA&gt; &lt;NA&gt;
## NA.47245   &lt;NA&gt;            &lt;NA&gt;     &lt;NA&gt;   NA         NA &lt;NA&gt; &lt;NA&gt;
## NA.47246   &lt;NA&gt;            &lt;NA&gt;     &lt;NA&gt;   NA         NA &lt;NA&gt; &lt;NA&gt;
## NA.47247   &lt;NA&gt;            &lt;NA&gt;     &lt;NA&gt;   NA         NA &lt;NA&gt; &lt;NA&gt;
## NA.47248   &lt;NA&gt;            &lt;NA&gt;     &lt;NA&gt;   NA         NA &lt;NA&gt; &lt;NA&gt;
## NA.47249   &lt;NA&gt;            &lt;NA&gt;     &lt;NA&gt;   NA         NA &lt;NA&gt; &lt;NA&gt;</code></pre>
<p>What is the problem with this approach? It appears that there are a bunch of <code>NA</code> values. Taking a quick look at the <code>symbol</code> column, we see what the problem.</p>
<pre class="r"><code>summary(ydat$symbol)</code></pre>
<pre><code>##      AAC1     AAD10     AAD14     AAD16      AAD4      AAD6      AAH1 
##        36        36        36        36        36        36        36 
##      AAP1     AAP1&#39;      AAT1      AAT2      ABC1      ABD1      ABF1 
##        36        36        36        36        36        36        36 
##      ABF2      ABP1    ABP140      ABZ1      ACA1      ACB1      ACC1 
##        36        36        36        36        36        36        36 
##      ACE2      ACF2      ACF4      ACH1      ACN9      ACO1      ACP1 
##        36        36        36        36        36        36        36 
##      ACS1      ACS2      ACT1      ADA2      ADE1     ADE12     ADE13 
##        36        36        36        36        36        36        36 
##     ADE16     ADE17      ADE2      ADE3      ADE4    ADE5,7      ADE6 
##        36        36        36        36        36        36        36 
##      ADE8      ADH1      ADH2      ADH3      ADH4      ADH5      ADH6 
##        36        36        36        36        36        36        36 
##      ADK1      ADO1      ADP1      ADR1      ADY3      ADY4      AEP1 
##        36        36        36        36        36        36        36 
##      AEP2      AEP3      AFG1      AFG2      AFG3      AFR1      AGA1 
##        36        36        36        36        36        36        36 
##      AGC1      AGE1      AGE2      AGP1      AGP3      AGX1      AHA1 
##        36        36        36        36        36        36        36 
##      AHC1      AHP1       AI1       AI2       AI4 AI5_ALPHA  AI5_BETA 
##        36        36        36        36        36        36        36 
##      AIP1      AIR1      AIR2      AKL1      AKR1      AKR2      ALA1 
##        36        36        36        36        36        36        36 
##      ALD2      ALD3      ALD4      ALD5      ALD6      ALF1      ALG1 
##        36        36        36        36        36        36        36 
##     ALG11     ALG12     ALG13     ALG14      ALG2      ALG3      ALG5 
##        36        36        36        36        36        36        36 
##   (Other)      NA&#39;s 
##    147652     47250</code></pre>
<p>Using the <code>is.na()</code> function, we can make filter further to get down to values of interest.</p>
<pre class="r"><code>head(ydat[ydat$symbol == &#39;LEU1&#39; &amp; !is.na(ydat$symbol), ])</code></pre>
<pre><code>##       symbol systematic_name nutrient rate expression                   bp
## 1526    LEU1         YGL009C  Glucose 0.05      -1.12 leucine biosynthesis
## 7043    LEU1         YGL009C  Glucose 0.10      -0.77 leucine biosynthesis
## 12555   LEU1         YGL009C  Glucose 0.15      -0.67 leucine biosynthesis
## 18071   LEU1         YGL009C  Glucose 0.20      -0.59 leucine biosynthesis
## 23603   LEU1         YGL009C  Glucose 0.25      -0.20 leucine biosynthesis
## 29136   LEU1         YGL009C  Glucose 0.30       0.03 leucine biosynthesis
##                                           mf
## 1526  3-isopropylmalate dehydratase activity
## 7043  3-isopropylmalate dehydratase activity
## 12555 3-isopropylmalate dehydratase activity
## 18071 3-isopropylmalate dehydratase activity
## 23603 3-isopropylmalate dehydratase activity
## 29136 3-isopropylmalate dehydratase activity</code></pre>
<p>Sometimes, looking at the data themselves is not that important. Using <code>dim()</code> is one possibility to look at the number of rows and columns after subsetting.</p>
<pre class="r"><code>dim(ydat[ydat$expression &gt; 3, ])</code></pre>
<pre><code>## [1] 714   7</code></pre>
<p>Find the high expressed genes when leucine-starved. For this task we can also use <code>subset</code> which allows us to treat column names as R variables (no <code>$</code> needed).</p>
<pre class="r"><code>subset(ydat, nutrient == &#39;Leucine&#39; &amp; rate == 0.05 &amp; expression &gt; 3)</code></pre>
<pre><code>##        symbol systematic_name nutrient rate expression
## 133768   QDR2         YIL121W  Leucine 0.05       4.61
## 133772   LEU1         YGL009C  Leucine 0.05       3.84
## 133858   BAP3         YDR046C  Leucine 0.05       4.29
## 135186   &lt;NA&gt;         YPL033C  Leucine 0.05       3.43
## 135187   &lt;NA&gt;         YLR267W  Leucine 0.05       3.23
## 135288   HXT3         YDR345C  Leucine 0.05       5.16
## 135963   TPO2         YGR138C  Leucine 0.05       3.75
## 135965   YRO2         YBR054W  Leucine 0.05       4.40
## 136102   GPG1         YGL121C  Leucine 0.05       3.08
## 136109  HSP42         YDR171W  Leucine 0.05       3.07
## 136119   HXT5         YHR096C  Leucine 0.05       4.90
## 136151   &lt;NA&gt;         YJL144W  Leucine 0.05       3.06
## 136152   MOH1         YBL049W  Leucine 0.05       3.43
## 136153   &lt;NA&gt;         YBL048W  Leucine 0.05       3.95
## 136189  HSP26         YBR072W  Leucine 0.05       4.86
## 136231   NCA3         YJL116C  Leucine 0.05       4.03
## 136233   &lt;NA&gt;         YBR116C  Leucine 0.05       3.28
## 136486   &lt;NA&gt;         YGR043C  Leucine 0.05       3.07
## 137443   ADH2         YMR303C  Leucine 0.05       4.15
## 137448   ICL1         YER065C  Leucine 0.05       3.54
## 137451   SFC1         YJR095W  Leucine 0.05       3.72
## 137569   MLS1         YNL117W  Leucine 0.05       3.76
##                                               bp
## 133768                       multidrug transport
## 133772                      leucine biosynthesis
## 133858                      amino acid transport
## 135186                                  meiosis*
## 135187                biological process unknown
## 135288                          hexose transport
## 135963                       polyamine transport
## 135965                biological process unknown
## 136102                       signal transduction
## 136109                       response to stress*
## 136119                          hexose transport
## 136151                   response to dessication
## 136152                biological process unknown
## 136153                                      &lt;NA&gt;
## 136189                       response to stress*
## 136231 mitochondrion organization and biogenesis
## 136233                                      &lt;NA&gt;
## 136486                biological process unknown
## 137443                             fermentation*
## 137448                          glyoxylate cycle
## 137451                       fumarate transport*
## 137569                          glyoxylate cycle
##                                            mf
## 133768         multidrug efflux pump activity
## 133772 3-isopropylmalate dehydratase activity
## 133858        amino acid transporter activity
## 135186             molecular function unknown
## 135187             molecular function unknown
## 135288          glucose transporter activity*
## 135963          spermine transporter activity
## 135965             molecular function unknown
## 136102             signal transducer activity
## 136109               unfolded protein binding
## 136119          glucose transporter activity*
## 136151             molecular function unknown
## 136152             molecular function unknown
## 136153                                   &lt;NA&gt;
## 136189               unfolded protein binding
## 136231             molecular function unknown
## 136233                                   &lt;NA&gt;
## 136486                 transaldolase activity
## 137443         alcohol dehydrogenase activity
## 137448              isocitrate lyase activity
## 137451 succinate:fumarate antiporter activity
## 137569               malate synthase activity</code></pre>
</div>
</div>
<div id="aggregating-data" class="section level2">
<h2>Aggregating data</h2>
<p>Aggregating data, or summarizing by category, is a common way to look for trends or differences in measurements between categories. Use <code>aggregate</code> to find the mean expression by gene symbol.</p>
<pre class="r"><code>head(aggregate(ydat$expression, by=list( ydat$symbol), mean))</code></pre>
<pre><code>##   Group.1           x
## 1    AAC1  0.52888889
## 2    AAC3 -0.21628571
## 3   AAD10  0.43833333
## 4   AAD14 -0.07166667
## 5   AAD16  0.24194444
## 6    AAD4 -0.79166667</code></pre>
<pre class="r"><code># or 
head(aggregate(expression ~ symbol, mean, data=ydat))</code></pre>
<pre><code>##   symbol  expression
## 1   AAC1  0.52888889
## 2   AAC3 -0.21628571
## 3  AAD10  0.43833333
## 4  AAD14 -0.07166667
## 5  AAD16  0.24194444
## 6   AAD4 -0.79166667</code></pre>
</div>
<div id="creating-a-data.frame-from-scratch" class="section level2">
<h2>Creating a data.frame from scratch</h2>
<p>Sometimes it is useful to combine related data into one object. For example, let’s simulate some data.</p>
<pre class="r"><code>smoker = factor(rep(c(&quot;smoker&quot;, &quot;non-smoker&quot;), each=50))
smoker_numeric = as.numeric(smoker)
x = rnorm(100)
risk = x + 2*smoker_numeric</code></pre>
<p>We have two varibles, <code>risk</code> and <code>smoker</code> that are related. We can make a data.frame out of them:</p>
<pre class="r"><code>smoker_risk = data.frame(smoker = smoker, risk = risk)
head(smoker_risk)</code></pre>
<pre><code>##   smoker     risk
## 1 smoker 3.203162
## 2 smoker 6.537133
## 3 smoker 3.877835
## 4 smoker 6.116651
## 5 smoker 4.823612
## 6 smoker 4.571653</code></pre>
<p>R also has plotting shortcuts that work with data.frames to simplify plotting</p>
<pre class="r"><code>plot( risk ~ smoker, data=smoker_risk)</code></pre>
<p><img src="dataframes_intro_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
</div>
<div id="saving-a-data.frame" class="section level2">
<h2>Saving a data.frame</h2>
<p>Once we have a data.frame of interest, we may want to save it. The most portable way to save a data.frame is to use one of the <code>write</code> functions. In this case, let’s save the data as a <code>.csv</code> file.</p>
<pre class="r"><code>write.csv(smoker_risk, &quot;smoker_risk.csv&quot;)</code></pre>
</div>

    <script>
        (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
        (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
        m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
        })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
        ga('create', 'UA-93043521-1', 'auto');
        ga('send', 'pageview');

         
        var links = document.querySelectorAll('a');
        Array.prototype.map.call(links, function(item) {
            if (item.host != document.location.host) {
                item.addEventListener('click', function() {
                    var action = item.getAttribute('data-action') || 'follow';
                    ga('send', 'event', 'outbound', action, item.href);
                });
            }
        });
    </script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
