<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Working with Genomic Ranges and Signals</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">seandavi(s12): Courses and Tutorials</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="setup.html">
    <span class="fa fa-cogs"></span>
     
    setup
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="ion ion-easel"></span>
     
    Slides
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="motivation_for_R_slides.html">Motivation for using R</a>
    </li>
    <li>
      <a href="http://bit.ly/bioc_cshl_2019">Introduction to Bioconductor</a>
    </li>
    <li>
      <a href="MachineLearning.html">Machine Learning hands-on</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-question fa-lg"></span>
     
    Misc.
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="further_resources.html">Further resources</a>
    </li>
    <li>
      <a href="https://github.com/seandavi/ITR">Source code for this site</a>
    </li>
    <li>
      <a href="https://github.com/seandavi/ITR/archive/master.zip">Download materials</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Working with Genomic Ranges and Signals</h1>
<h4 class="author">Sean Davis</h4>
<address class="author_afil">
Center for Cancer Research, National Cancer Institute, NIH<br><a class="author_email" href="mailto:#"><a href="mailto:seandavi@gmail.com">seandavi@gmail.com</a></a>
</address>

</div>


<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<div id="getting-started" class="section level1">
<h1>Getting started</h1>
<p>Genomic ranges are a way of describing regions on the genome (or any other linear object, such as a transcript, or even a protein). This functionality is typically found in the <em><a href="https://bioconductor.org/packages/3.10/GenomicRanges">GenomicRanges</a></em> package.</p>
<pre class="r"><code>library(GenomicRanges)</code></pre>
<div id="example-granges" class="section level2">
<h2>Example GRanges</h2>
<p>To get going, we can construct a <code>GRanges</code> object by hand as an example.</p>
<pre class="r"><code>gr &lt;- GRanges(
    seqnames = Rle(c(&quot;chr1&quot;, &quot;chr2&quot;, &quot;chr1&quot;, &quot;chr3&quot;), c(1, 3, 2, 4)),
    ranges = IRanges(101:110, end = 111:120, names = head(letters, 10)),
    strand = Rle(strand(c(&quot;-&quot;, &quot;+&quot;, &quot;*&quot;, &quot;+&quot;, &quot;-&quot;)), c(1, 2, 2, 3, 2)),
    score = 1:10,
    GC = seq(1, 0, length=10))
gr</code></pre>
<pre><code>## GRanges object with 10 ranges and 2 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   a     chr1   101-111      - |         1                 1
##   b     chr2   102-112      + |         2 0.888888888888889
##   c     chr2   103-113      + |         3 0.777777777777778
##   d     chr2   104-114      * |         4 0.666666666666667
##   e     chr1   105-115      * |         5 0.555555555555556
##   f     chr1   106-116      + |         6 0.444444444444444
##   g     chr3   107-117      + |         7 0.333333333333333
##   h     chr3   108-118      + |         8 0.222222222222222
##   i     chr3   109-119      - |         9 0.111111111111111
##   j     chr3   110-120      - |        10                 0
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>The <code>GRanges</code> class represents a collection of genomic ranges that each have a single start and end location on the genome. It can be used to store the location of genomic features such as contiguous binding sites, transcripts, and exons. These objects can be created by using the GRanges constructor function. The following code just creates a <code>GRanges</code> object from scratch.</p>
<p>This creates a <code>GRanges</code> object with 10 genomic ranges. The output of the <code>GRanges</code> <code>show()</code> method separates the information into a left and right hand region that are separated by <code>|</code> symbols. The genomic coordinates (seqnames, ranges, and strand) are located on the left-hand side and the metadata columns are located on the right. For this example, the metadata is comprised of score and GC information, but almost anything can be stored in the metadata portion of a GRanges object.</p>
<p>The components of the genomic coordinates within a GRanges object can be extracted using the seqnames, ranges, and strand accessor functions.</p>
<pre class="r"><code>seqnames(gr)</code></pre>
<pre><code>## factor-Rle of length 10 with 4 runs
##   Lengths:    1    3    2    4
##   Values : chr1 chr2 chr1 chr3
## Levels(3): chr1 chr2 chr3</code></pre>
<pre class="r"><code>ranges(gr)</code></pre>
<pre><code>## IRanges object with 10 ranges and 0 metadata columns:
##         start       end     width
##     &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   a       101       111        11
##   b       102       112        11
##   c       103       113        11
##   d       104       114        11
##   e       105       115        11
##   f       106       116        11
##   g       107       117        11
##   h       108       118        11
##   i       109       119        11
##   j       110       120        11</code></pre>
<pre class="r"><code>strand(gr)</code></pre>
<pre><code>## factor-Rle of length 10 with 5 runs
##   Lengths: 1 2 2 3 2
##   Values : - + * + -
## Levels(3): + - *</code></pre>
<p>Note that the <code>GRanges</code> object has information to the “left” side of the <code>|</code> that has special accessors. The information to the right side of the <code>|</code>, when it is present, is the metadata and is accessed using <code>mcols()</code>, for “metadata columns”.</p>
<pre class="r"><code>class(mcols(gr))</code></pre>
<pre><code>## [1] &quot;DataFrame&quot;
## attr(,&quot;package&quot;)
## [1] &quot;S4Vectors&quot;</code></pre>
<pre class="r"><code>mcols(gr)</code></pre>
<pre><code>## DataFrame with 10 rows and 2 columns
##       score                GC
##   &lt;integer&gt;         &lt;numeric&gt;
## a         1                 1
## b         2 0.888888888888889
## c         3 0.777777777777778
## d         4 0.666666666666667
## e         5 0.555555555555556
## f         6 0.444444444444444
## g         7 0.333333333333333
## h         8 0.222222222222222
## i         9 0.111111111111111
## j        10                 0</code></pre>
<p>Since the <code>class</code> of <code>mcols(gr)</code> is DataFrame, we can use our <code>DataFrame</code> approaches to work with the data.</p>
<pre class="r"><code>mcols(gr)$score</code></pre>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10</code></pre>
<p>We can even assign a new column.</p>
<pre class="r"><code>mcols(gr)$AT = 1-mcols(gr)$GC
gr</code></pre>
<pre><code>## GRanges object with 10 ranges and 3 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   a     chr1   101-111      - |         1                 1
##   b     chr2   102-112      + |         2 0.888888888888889
##   c     chr2   103-113      + |         3 0.777777777777778
##   d     chr2   104-114      * |         4 0.666666666666667
##   e     chr1   105-115      * |         5 0.555555555555556
##   f     chr1   106-116      + |         6 0.444444444444444
##   g     chr3   107-117      + |         7 0.333333333333333
##   h     chr3   108-118      + |         8 0.222222222222222
##   i     chr3   109-119      - |         9 0.111111111111111
##   j     chr3   110-120      - |        10                 0
##                    AT
##             &lt;numeric&gt;
##   a                 0
##   b 0.111111111111111
##   c 0.222222222222222
##   d 0.333333333333333
##   e 0.444444444444444
##   f 0.555555555555556
##   g 0.666666666666667
##   h 0.777777777777778
##   i 0.888888888888889
##   j                 1
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>Another common way to create a <code>GRanges</code> object is to start with a <code>data.frame</code>, perhaps created by hand like below or read in using <code>read.csv</code> or <code>read.table</code>. We can convert from a <code>data.frame</code>, when columns are named appropriately, to a <code>GRanges</code> object.</p>
<pre class="r"><code>df_regions = data.frame(chromosome = rep(&quot;chr1&quot;,10),
                        start=seq(1000,10000,1000),
                        end=seq(1100, 10100, 1000))
as(df_regions,&#39;GRanges&#39;) # note that names have to match with GRanges slots</code></pre>
<pre><code>## GRanges object with 10 ranges and 0 metadata columns:
##        seqnames      ranges strand
##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     chr1   1000-1100      *
##    [2]     chr1   2000-2100      *
##    [3]     chr1   3000-3100      *
##    [4]     chr1   4000-4100      *
##    [5]     chr1   5000-5100      *
##    [6]     chr1   6000-6100      *
##    [7]     chr1   7000-7100      *
##    [8]     chr1   8000-8100      *
##    [9]     chr1   9000-9100      *
##   [10]     chr1 10000-10100      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code># fix column name
colnames(df_regions)[1] = &#39;seqnames&#39;
gr2 = as(df_regions,&#39;GRanges&#39;)
gr2</code></pre>
<pre><code>## GRanges object with 10 ranges and 0 metadata columns:
##        seqnames      ranges strand
##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     chr1   1000-1100      *
##    [2]     chr1   2000-2100      *
##    [3]     chr1   3000-3100      *
##    [4]     chr1   4000-4100      *
##    [5]     chr1   5000-5100      *
##    [6]     chr1   6000-6100      *
##    [7]     chr1   7000-7100      *
##    [8]     chr1   8000-8100      *
##    [9]     chr1   9000-9100      *
##   [10]     chr1 10000-10100      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p><code>GRanges</code> have one-dimensional-like behavior. For instance, we can check the <code>length</code> and even give <code>GRanges</code> names.</p>
<pre class="r"><code>names(gr)</code></pre>
<pre><code>##  [1] &quot;a&quot; &quot;b&quot; &quot;c&quot; &quot;d&quot; &quot;e&quot; &quot;f&quot; &quot;g&quot; &quot;h&quot; &quot;i&quot; &quot;j&quot;</code></pre>
<pre class="r"><code>length(gr)</code></pre>
<pre><code>## [1] 10</code></pre>
</div>
<div id="subsetting-granges-objects" class="section level2">
<h2>Subsetting GRanges objects</h2>
<p>While <code>GRanges</code> objects look a bit like a <code>data.frame</code>, they can be thought of as vectors with associated ranges. Subsetting, then, works very similarly to vectors. To subset a <code>GRanges</code> object to include only second and third regions:</p>
<pre class="r"><code>gr[2:3]</code></pre>
<pre><code>## GRanges object with 2 ranges and 3 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   b     chr2   102-112      + |         2 0.888888888888889
##   c     chr2   103-113      + |         3 0.777777777777778
##                    AT
##             &lt;numeric&gt;
##   b 0.111111111111111
##   c 0.222222222222222
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>That said, if the <code>GRanges</code> object has metadata columns, we can also treat it like a two-dimensional object kind of like a data.frame. Note that the information to the left of the <code>|</code> is not like a <code>data.frame</code>, so we <em>cannot</em> do something like <code>gr$seqnames</code>. Here is an example of subsetting with the subset of one metadata column.</p>
<pre class="r"><code>gr[2:3, &quot;GC&quot;]</code></pre>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##     seqnames    ranges strand |                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   b     chr2   102-112      + | 0.888888888888889
##   c     chr2   103-113      + | 0.777777777777778
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>The usual <code>head()</code> and <code>tail()</code> also work just fine.</p>
<pre class="r"><code>head(gr,n=2)</code></pre>
<pre><code>## GRanges object with 2 ranges and 3 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   a     chr1   101-111      - |         1                 1
##   b     chr2   102-112      + |         2 0.888888888888889
##                    AT
##             &lt;numeric&gt;
##   a                 0
##   b 0.111111111111111
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>tail(gr,n=2)</code></pre>
<pre><code>## GRanges object with 2 ranges and 3 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   i     chr3   109-119      - |         9 0.111111111111111
##   j     chr3   110-120      - |        10                 0
##                    AT
##             &lt;numeric&gt;
##   i 0.888888888888889
##   j                 1
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="interval-operations-on-one-granges-object" class="section level2">
<h2>Interval operations on one GRanges object</h2>
<div id="intra-range-methods" class="section level3">
<h3>Intra-range methods</h3>
<p>The methods described in this section work <em>one-region-at-a-time</em> and are, therefore, called “intra-region” methods. Methods that work across all regions are described below in the <a href="#inter-range-methods">Inter-range methods</a> section.</p>
<p>The <code>GRanges</code> class has accessors for the “ranges” part of the data. For example:</p>
<pre class="r"><code># Make a smaller GRanges subset
g &lt;- gr[1:3]
start(g) # to get start locations</code></pre>
<pre><code>## [1] 101 102 103</code></pre>
<pre class="r"><code>end(g)   # to get end locations</code></pre>
<pre><code>## [1] 111 112 113</code></pre>
<pre class="r"><code>width(g) # to get the &quot;widths&quot; of each range</code></pre>
<pre><code>## [1] 11 11 11</code></pre>
<pre class="r"><code>range(g) # to get the &quot;range&quot; for each sequence (min(start) through max(end))</code></pre>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1   101-111      -
##   [2]     chr2   102-113      +
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>The GRanges class also has many methods for manipulating the ranges. The methods can be classified as intra-range methods, inter-range methods, and between-range methods. Intra-range methods operate on each element of a GRanges object independent of the other ranges in the object. For example, the flank method can be used to recover regions flanking the set of ranges represented by the GRanges object. So to get a GRanges object containing the ranges that include the 10 bases <em>upstream</em> of the ranges:</p>
<pre class="r"><code>flank(g, 10)</code></pre>
<pre><code>## GRanges object with 3 ranges and 3 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   a     chr1   112-121      - |         1                 1
##   b     chr2    92-101      + |         2 0.888888888888889
##   c     chr2    93-102      + |         3 0.777777777777778
##                    AT
##             &lt;numeric&gt;
##   a                 0
##   b 0.111111111111111
##   c 0.222222222222222
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>Note how <code>flank</code> pays attention to “strand”. To get the flanking regions <em>downstream</em> of the ranges, we can do:</p>
<pre class="r"><code>flank(g, 10, start=FALSE)</code></pre>
<pre><code>## GRanges object with 3 ranges and 3 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   a     chr1    91-100      - |         1                 1
##   b     chr2   113-122      + |         2 0.888888888888889
##   c     chr2   114-123      + |         3 0.777777777777778
##                    AT
##             &lt;numeric&gt;
##   a                 0
##   b 0.111111111111111
##   c 0.222222222222222
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>Other examples of intra-range methods include <code>resize</code> and <code>shift</code>. The shift method will move the ranges by a specific number of base pairs, and the resize method will extend the ranges by a specified width.</p>
<pre class="r"><code>shift(g, 5)</code></pre>
<pre><code>## GRanges object with 3 ranges and 3 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   a     chr1   106-116      - |         1                 1
##   b     chr2   107-117      + |         2 0.888888888888889
##   c     chr2   108-118      + |         3 0.777777777777778
##                    AT
##             &lt;numeric&gt;
##   a                 0
##   b 0.111111111111111
##   c 0.222222222222222
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>resize(g, 30)</code></pre>
<pre><code>## GRanges object with 3 ranges and 3 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   a     chr1    82-111      - |         1                 1
##   b     chr2   102-131      + |         2 0.888888888888889
##   c     chr2   103-132      + |         3 0.777777777777778
##                    AT
##             &lt;numeric&gt;
##   a                 0
##   b 0.111111111111111
##   c 0.222222222222222
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>The <em><a href="https://bioconductor.org/packages/3.10/GenomicRanges">GenomicRanges</a></em> help page <code>?&quot;intra-range-methods&quot;</code> summarizes these methods.</p>
</div>
<div id="inter-range-methods" class="section level3">
<h3>Inter-range methods</h3>
<p>Inter-range methods involve comparisons between ranges in a single GRanges object. For instance, the reduce method will align the ranges and merge overlapping ranges to produce a simplified set.</p>
<pre class="r"><code>reduce(g)</code></pre>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1   101-111      -
##   [2]     chr2   102-113      +
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>The reduce method could, for example, be used to collapse individual overlapping coding exons into a single set of coding regions.</p>
<p>Sometimes one is interested in the gaps or the qualities of the gaps between the ranges represented by your GRanges object. The gaps method provides this information:</p>
<pre class="r"><code>gaps(g)</code></pre>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1     1-100      -
##   [2]     chr2     1-101      +
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>In this case, we have not specified the lengths of the chromosomes, so Bioconductor is making the assumption (incorrectly) that the chromosomes end at the largest location on each chromosome. We can correct this by setting the <code>seqlengths</code> correctly, but we can ignore that detail for now.</p>
<p>The disjoin method represents a GRanges object as a collection of non-overlapping ranges:</p>
<pre class="r"><code>disjoin(g)</code></pre>
<pre><code>## GRanges object with 4 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1   101-111      -
##   [2]     chr2       102      +
##   [3]     chr2   103-112      +
##   [4]     chr2       113      +
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>The <code>coverage</code> method quantifies the degree of overlap for all the ranges in a GRanges object.</p>
<pre class="r"><code>coverage(g)</code></pre>
<pre><code>## RleList of length 3
## $chr1
## integer-Rle of length 111 with 2 runs
##   Lengths: 100  11
##   Values :   0   1
## 
## $chr2
## integer-Rle of length 113 with 4 runs
##   Lengths: 101   1  10   1
##   Values :   0   1   2   1
## 
## $chr3
## integer-Rle of length 0 with 0 runs
##   Lengths: 
##   Values :</code></pre>
<p>The coverage is summarized as a <code>list</code> of coverages, one for each chromosome. The <code>Rle</code> class is used to store the values. Sometimes, one must convert these values to <code>numeric</code> using <code>as.numeric</code>. In many cases, this will happen automatically, though.</p>
<pre class="r"><code>covg = coverage(g)
covg_chr2 = covg[[&#39;chr2&#39;]]
plot(covg_chr2, type=&#39;l&#39;)</code></pre>
<p><img src="RangesAndSignal_files/figure-html/coveragegr2-1.png" width="672" /></p>
<p>See the <em><a href="https://bioconductor.org/packages/3.10/GenomicRanges">GenomicRanges</a></em> help page <code>?&quot;intra-range-methods&quot;</code> for more details.</p>
</div>
</div>
<div id="set-operations-for-granges-objects" class="section level2">
<h2>Set operations for GRanges objects</h2>
<p>Between-range methods calculate relationships between different GRanges objects. Of central importance are findOverlaps and related operations; these are discussed below. Additional operations treat GRanges as mathematical sets of coordinates; union(g, g2) is the union of the coordinates in g and g2. Here are examples for calculating the union, the intersect and the asymmetric difference (using setdiff).</p>
<pre class="r"><code>g2 &lt;- head(gr, n=2)
union(g, g2)</code></pre>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1   101-111      -
##   [2]     chr2   102-113      +
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>intersect(g, g2)</code></pre>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1   101-111      -
##   [2]     chr2   102-112      +
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>setdiff(g, g2)</code></pre>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr2       113      +
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>There is extensive additional help available or by looking at the vignettes in at the <em><a href="https://bioconductor.org/packages/3.10/GenomicRanges">GenomicRanges</a></em> pages.</p>
<pre class="r"><code>?GRanges</code></pre>
<p>There are also many possible <code>methods</code> that work with <code>GRanges</code> objects. To see a complete list (long), try:</p>
<pre class="r"><code>methods(class=&quot;GRanges&quot;)</code></pre>
</div>
</div>
<div id="grangeslist" class="section level1">
<h1>GRangesList</h1>
<p>Some important genomic features, such as spliced transcripts that are are comprised of exons, are inherently compound structures. Such a feature makes much more sense when expressed as a compound object such as a GRangesList. If we thing of each transcript as a set of exons, each transcript would be summarized as a <code>GRanges</code> object. However, if we have multiple transcripts, we want to somehow keep them separate, with each transcript having its own exons. The <code>GRangesList</code> is then a list of <code>GRanges</code> objects that. Continuing with the transcripts thought, a <code>GRangesList</code> can contain all the transcripts and their exons; each transcript is an element in the list.</p>
<p>Whenever genomic features consist of multiple ranges that are grouped by a parent feature, they can be represented as a GRangesList object. Consider the simple example of the two transcript GRangesList below created using the GRangesList constructor.</p>
<pre class="r"><code>gr1 &lt;- GRanges(
    seqnames = &quot;chr2&quot;, 
    ranges = IRanges(103, 106),
    strand = &quot;+&quot;, 
    score = 5L, GC = 0.45)
gr2 &lt;- GRanges(
    seqnames = c(&quot;chr1&quot;, &quot;chr1&quot;),
    ranges = IRanges(c(107, 113), width = 3),
    strand = c(&quot;+&quot;, &quot;-&quot;), 
    score = 3:4, GC = c(0.3, 0.5))</code></pre>
<p>The <code>gr1</code> and <code>gr2</code> are each <code>GRanges</code> objects. We can combine them into a “named” <code>GRangesList</code> like so:</p>
<pre class="r"><code>grl &lt;- GRangesList(&quot;txA&quot; = gr1, &quot;txB&quot; = gr2)
grl</code></pre>
<pre><code>## GRangesList object of length 2:
## $txA
## GRanges object with 1 range and 2 metadata columns:
##       seqnames    ranges strand |     score        GC
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
##   [1]     chr2   103-106      + |         5      0.45
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
## 
## $txB
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames    ranges strand |     score        GC
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
##   [1]     chr1   107-109      + |         3       0.3
##   [2]     chr1   113-115      - |         4       0.5
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>The <code>show</code> method for a <code>GRangesList</code> object displays it as a named list of <code>GRanges</code> objects, where the names of this list are considered to be the names of the grouping feature. In the example above, the groups of individual exon ranges are represented as separate <code>GRanges</code> objects which are further organized into a list structure where each element name is a transcript name. Many other combinations of grouped and labeled <code>GRanges</code> objects are possible of course, but this example is a common arrangement.</p>
<p>In some cases, <code>GRangesList</code>s behave quite similarly to <code>GRanges</code> objects.</p>
<div id="basic-grangeslist-accessors" class="section level2">
<h2>Basic <em>GRangesList</em> accessors</h2>
<p>Just as with GRanges object, the components of the genomic coordinates within a GRangesList object can be extracted using simple accessor methods. Not surprisingly, the GRangesList objects have many of the same accessors as GRanges objects. The difference is that many of these methods return a list since the input is now essentially a list of GRanges objects. Here are a few examples:</p>
<pre class="r"><code>seqnames(grl)</code></pre>
<pre><code>## RleList of length 2
## $txA
## factor-Rle of length 1 with 1 run
##   Lengths:    1
##   Values : chr2
## Levels(2): chr2 chr1
## 
## $txB
## factor-Rle of length 2 with 1 run
##   Lengths:    2
##   Values : chr1
## Levels(2): chr2 chr1</code></pre>
<pre class="r"><code>ranges(grl)</code></pre>
<pre><code>## IRangesList object of length 2:
## $txA
## IRanges object with 1 range and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]       103       106         4
## 
## $txB
## IRanges object with 2 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]       107       109         3
##   [2]       113       115         3</code></pre>
<pre class="r"><code>strand(grl)</code></pre>
<pre><code>## RleList of length 2
## $txA
## factor-Rle of length 1 with 1 run
##   Lengths: 1
##   Values : +
## Levels(3): + - *
## 
## $txB
## factor-Rle of length 2 with 2 runs
##   Lengths: 1 1
##   Values : + -
## Levels(3): + - *</code></pre>
<p>The length and names methods will return the length or names of the list and the seqlengths method will return the set of sequence lengths.</p>
<pre class="r"><code>length(grl)</code></pre>
<pre><code>## [1] 2</code></pre>
<pre class="r"><code>names(grl)</code></pre>
<pre><code>## [1] &quot;txA&quot; &quot;txB&quot;</code></pre>
<pre class="r"><code>seqlengths(grl)</code></pre>
<pre><code>## chr2 chr1 
##   NA   NA</code></pre>
</div>
</div>
<div id="relationships-between-region-sets" class="section level1">
<h1>Relationships between region sets</h1>
<p>One of the more powerful approaches to genomic data integration is to ask about the relationship between sets of genomic ranges. The key features of this process are to look at overlaps and distances to the nearest feature. These functionalities, combined with the operations like <code>flank</code> and <code>resize</code>, for instance, allow pretty useful analyses with relatively little code. In general, these operations are <em>very</em> fast, even on thousands to millions of regions.</p>
<div id="overlaps" class="section level2">
<h2>Overlaps</h2>
<pre class="r"><code>mtch &lt;- findOverlaps(gr, g)
as.matrix(mtch)</code></pre>
<pre><code>##      queryHits subjectHits
## [1,]         1           1
## [2,]         2           2
## [3,]         2           3
## [4,]         3           2
## [5,]         3           3
## [6,]         4           2
## [7,]         4           3
## [8,]         5           1</code></pre>
<p>If you are interested in only the <code>queryHits</code> or the <code>subjectHits</code>, there are accessors for those (ie., <code>queryHits(match)</code>).</p>
<p>As you might expect, the <code>countOverlaps</code> method counts the regions in the second <code>GRanges</code> that overlap with those that overlap with each element of the first.</p>
<pre class="r"><code>countOverlaps(gr, g)</code></pre>
<pre><code>## a b c d e f g h i j 
## 1 2 2 2 1 0 0 0 0 0</code></pre>
<p>The <code>subsetByOverlaps</code> method simply subsets the first <code>GRanges</code> object to include <em>only</em> those that overlap the second.</p>
<pre class="r"><code>subsetByOverlaps(gr, g)</code></pre>
<pre><code>## GRanges object with 5 ranges and 3 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   a     chr1   101-111      - |         1                 1
##   b     chr2   102-112      + |         2 0.888888888888889
##   c     chr2   103-113      + |         3 0.777777777777778
##   d     chr2   104-114      * |         4 0.666666666666667
##   e     chr1   105-115      * |         5 0.555555555555556
##                    AT
##             &lt;numeric&gt;
##   a                 0
##   b 0.111111111111111
##   c 0.222222222222222
##   d 0.333333333333333
##   e 0.444444444444444
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>In some cases, you may be interested in only one hit when doing overlaps. Note the <code>select</code> parameter. See the help for <code>findOverlaps</code></p>
<pre class="r"><code>findOverlaps(gr, g, select=&quot;first&quot;)</code></pre>
<pre><code>##  [1]  1  2  2  2  1 NA NA NA NA NA</code></pre>
<pre class="r"><code>findOverlaps(g, gr, select=&quot;first&quot;)</code></pre>
<pre><code>## [1] 1 2 2</code></pre>
<p>The <code>%over%</code> logical operator allows us to do similar things to <code>findOverlaps</code> and <code>subsetByOverlaps</code>.</p>
<pre class="r"><code>gr %over% g</code></pre>
<pre><code>##  [1]  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE</code></pre>
<pre class="r"><code>gr[gr %over% g]</code></pre>
<pre><code>## GRanges object with 5 ranges and 3 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   a     chr1   101-111      - |         1                 1
##   b     chr2   102-112      + |         2 0.888888888888889
##   c     chr2   103-113      + |         3 0.777777777777778
##   d     chr2   104-114      * |         4 0.666666666666667
##   e     chr1   105-115      * |         5 0.555555555555556
##                    AT
##             &lt;numeric&gt;
##   a                 0
##   b 0.111111111111111
##   c 0.222222222222222
##   d 0.333333333333333
##   e 0.444444444444444
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="nearest-feature" class="section level2">
<h2>Nearest feature</h2>
<p>There are a number of useful methods that find the nearest feature (region) in a second set for each element in the first set.</p>
<p>We can review our two <code>GRanges</code> toy objects:</p>
<pre class="r"><code>g</code></pre>
<pre><code>## GRanges object with 3 ranges and 3 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   a     chr1   101-111      - |         1                 1
##   b     chr2   102-112      + |         2 0.888888888888889
##   c     chr2   103-113      + |         3 0.777777777777778
##                    AT
##             &lt;numeric&gt;
##   a                 0
##   b 0.111111111111111
##   c 0.222222222222222
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>gr</code></pre>
<pre><code>## GRanges object with 10 ranges and 3 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   a     chr1   101-111      - |         1                 1
##   b     chr2   102-112      + |         2 0.888888888888889
##   c     chr2   103-113      + |         3 0.777777777777778
##   d     chr2   104-114      * |         4 0.666666666666667
##   e     chr1   105-115      * |         5 0.555555555555556
##   f     chr1   106-116      + |         6 0.444444444444444
##   g     chr3   107-117      + |         7 0.333333333333333
##   h     chr3   108-118      + |         8 0.222222222222222
##   i     chr3   109-119      - |         9 0.111111111111111
##   j     chr3   110-120      - |        10                 0
##                    AT
##             &lt;numeric&gt;
##   a                 0
##   b 0.111111111111111
##   c 0.222222222222222
##   d 0.333333333333333
##   e 0.444444444444444
##   f 0.555555555555556
##   g 0.666666666666667
##   h 0.777777777777778
##   i 0.888888888888889
##   j                 1
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<ul>
<li><p>nearest: Performs conventional nearest neighbor finding. Returns an integer vector containing the index of the nearest neighbor range in subject for each range in x. If there is no nearest neighbor NA is returned. For details of the algorithm see the man page in the IRanges package (?nearest).</p></li>
<li><p>precede: For each range in x, precede returns the index of the range in subject that is directly preceded by the range in x. Overlapping ranges are excluded. NA is returned when there are no qualifying ranges in subject.</p></li>
<li><p>follow: The opposite of precede, follow returns the index of the range in subject that is directly followed by the range in x. Overlapping ranges are excluded. NA is returned when there are no qualifying ranges in subject.</p></li>
</ul>
<p>Orientation and strand for precede and follow: Orientation is 5’ to 3’, consistent with the direction of translation. Because positional numbering along a chromosome is from left to right and transcription takes place from 5’ to 3’, precede and follow can appear to have ‘opposite’ behavior on the + and - strand. Using positions 5 and 6 as an example, 5 precedes 6 on the + strand but follows 6 on the - strand.</p>
<p>The table below outlines the orientation when ranges on different strands are compared. In general, a feature on * is considered to belong to both strands. The single exception is when both x and subject are * in which case both are treated as +.</p>
<pre><code>       x  |  subject  |  orientation 
     -----+-----------+----------------
a)     +  |  +        |  ---&gt; 
b)     +  |  -        |  NA
c)     +  |  *        |  ---&gt;
d)     -  |  +        |  NA
e)     -  |  -        |  &lt;---
f)     -  |  *        |  &lt;---
g)     *  |  +        |  ---&gt;
h)     *  |  -        |  &lt;---
i)     *  |  *        |  ---&gt;  (the only situation where * arbitrarily means +)</code></pre>
<pre class="r"><code>res = nearest(g, gr)
res</code></pre>
<pre><code>## [1] 5 4 4</code></pre>
<p>While <code>nearest</code> and friends give the index of the nearest feature, the distance to the nearest is sometimes also useful to have. The <code>distanceToNearest</code> method calculates the nearest feature as well as reporting the distance.</p>
<pre class="r"><code>res = distanceToNearest(g, gr)
res</code></pre>
<pre><code>## Hits object with 3 hits and 1 metadata column:
##       queryHits subjectHits |  distance
##       &lt;integer&gt;   &lt;integer&gt; | &lt;integer&gt;
##   [1]         1           5 |         0
##   [2]         2           4 |         0
##   [3]         3           4 |         0
##   -------
##   queryLength: 3 / subjectLength: 10</code></pre>
</div>
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R Under development (unstable) (2019-01-14 r75992)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Mojave 10.14.2
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2
##  [2] RColorBrewer_1.1-2                     
##  [3] GEOquery_2.53.0                        
##  [4] tximport_1.13.6                        
##  [5] GenomicFeatures_1.37.3                 
##  [6] GSE5859_1.0                            
##  [7] tissuesGeneExpression_1.0              
##  [8] dplyr_0.8.1                            
##  [9] rafalib_1.0.0                          
## [10] usethis_1.5.0                          
## [11] devtools_2.0.2                         
## [12] airway_1.5.0                           
## [13] SummarizedExperiment_1.15.5            
## [14] DelayedArray_0.11.2                    
## [15] BiocParallel_1.19.0                    
## [16] matrixStats_0.54.0                     
## [17] nycflights13_1.0.0                     
## [18] ggplot2_3.2.0                          
## [19] knitr_1.23                             
## [20] BSgenome.Hsapiens.UCSC.hg19_1.4.0      
## [21] BSgenome_1.53.0                        
## [22] Biostrings_2.53.0                      
## [23] XVector_0.25.0                         
## [24] AnnotationDbi_1.47.0                   
## [25] Biobase_2.45.0                         
## [26] rtracklayer_1.45.1                     
## [27] GenomicRanges_1.37.14                  
## [28] GenomeInfoDb_1.21.1                    
## [29] IRanges_2.19.10                        
## [30] S4Vectors_0.23.17                      
## [31] AnnotationHub_2.17.3                   
## [32] BiocFileCache_1.9.1                    
## [33] dbplyr_1.4.2                           
## [34] BiocGenerics_0.31.4                    
## [35] BiocStyle_2.13.2                       
## 
## loaded via a namespace (and not attached):
##  [1] colorspace_1.4-1              rprojroot_1.3-2              
##  [3] revealjs_0.9                  fs_1.3.1                     
##  [5] remotes_2.1.0                 bit64_0.9-7                  
##  [7] interactiveDisplayBase_1.23.0 fansi_0.4.0                  
##  [9] xml2_1.2.0                    codetools_0.2-16             
## [11] pkgload_1.0.2                 zeallot_0.1.0                
## [13] Rsamtools_2.1.2               shiny_1.3.2                  
## [15] BiocManager_1.30.4            readr_1.3.1                  
## [17] compiler_3.6.0                httr_1.4.0                   
## [19] backports_1.1.4               assertthat_0.2.1             
## [21] Matrix_1.2-17                 lazyeval_0.2.2               
## [23] limma_3.41.6                  cli_1.1.0                    
## [25] later_0.8.0                   htmltools_0.3.6              
## [27] prettyunits_1.0.2             tools_3.6.0                  
## [29] gtable_0.3.0                  glue_1.3.1                   
## [31] GenomeInfoDbData_1.2.1        rappdirs_0.3.1               
## [33] Rcpp_1.0.1                    vctrs_0.1.0                  
## [35] xfun_0.8                      stringr_1.4.0                
## [37] ps_1.3.0                      testthat_2.1.1               
## [39] mime_0.7                      XML_3.98-1.20                
## [41] zlibbioc_1.31.0               MASS_7.3-51.4                
## [43] scales_1.0.0                  hms_0.4.2                    
## [45] promises_1.0.1                yaml_2.2.0                   
## [47] curl_3.3                      memoise_1.1.0                
## [49] biomaRt_2.41.3                stringi_1.4.3                
## [51] RSQLite_2.1.1                 highr_0.8                    
## [53] desc_1.2.0                    pkgbuild_1.0.3               
## [55] rlang_0.4.0                   pkgconfig_2.0.2              
## [57] bitops_1.0-6                  evaluate_0.14                
## [59] lattice_0.20-38               purrr_0.3.2                  
## [61] GenomicAlignments_1.21.4      labeling_0.3                 
## [63] bit_1.1-14                    processx_3.3.1               
## [65] tidyselect_0.2.5              magrittr_1.5                 
## [67] R6_2.4.0                      DBI_1.0.0                    
## [69] pillar_1.4.2                  withr_2.1.2                  
## [71] RCurl_1.95-4.12               tibble_2.1.3                 
## [73] crayon_1.3.4                  utf8_1.1.4                   
## [75] rmarkdown_1.13                progress_1.2.2               
## [77] grid_3.6.0                    blob_1.1.1                   
## [79] callr_3.2.0                   digest_0.6.19                
## [81] xtable_1.8-4                  tidyr_0.8.3                  
## [83] httpuv_1.5.1                  munsell_0.5.0                
## [85] sessioninfo_1.1.1</code></pre>
</div>

    <script>
        (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
        (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
        m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
        })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
        ga('create', 'UA-93043521-1', 'auto');
        ga('send', 'pageview');

         
        var links = document.querySelectorAll('a');
        Array.prototype.map.call(links, function(item) {
            if (item.host != document.location.host) {
                item.addEventListener('click', function() {
                    var action = item.getAttribute('data-action') || 'follow';
                    ga('send', 'event', 'outbound', action, item.href);
                });
            }
        });
    </script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
