<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Stephen Turner (with modifications by Sean Davis)" />


<title>Dataframes with dplyr</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">seandavi(s12): Courses and Tutorials</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="setup.html">
    <span class="fa fa-cogs"></span>
     
    setup
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="ion ion-easel"></span>
     
    Slides
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="motivation_for_R_slides.html">Motivation for using R</a>
    </li>
    <li>
      <a href="http://bit.ly/bioc_cshl_2019">Introduction to Bioconductor</a>
    </li>
    <li>
      <a href="https://drive.google.com/file/d/1txUz-a84VVxiB1ouv24ujL2DSTfxgblL/view?usp=sharing">Advanced Bioconductor Overview</a>
    </li>
    <li>
      <a href="MachineLearning.html">Machine Learning hands-on</a>
    </li>
    <li>
      <a href="https://docs.google.com/presentation/d/1PKP39ze3kATKCXxx-AUuDdI4FUpA85UQJxDMhXIK3Mk/edit?usp=sharing">Machine Learning Intro</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-question fa-lg"></span>
     
    Misc.
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="further_resources.html">Further resources</a>
    </li>
    <li>
      <a href="https://github.com/seandavi/ITR">Source code for this site</a>
    </li>
    <li>
      <a href="https://github.com/seandavi/ITR/archive/master.zip">Download materials</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Dataframes with dplyr</h1>
<h4 class="author">Stephen Turner (with modifications by Sean
Davis)</h4>

</div>


<p>Data analysis involves a large amount of <a
href="http://www.nytimes.com/2014/08/18/technology/for-big-data-scientists-hurdle-to-insights-is-janitor-work.html">janitor
work</a> – munging and cleaning data to facilitate downstream data
analysis. This lesson demonstrates techniques for advanced data
manipulation and analysis with the split-apply-combine strategy. We will
use the dplyr package in R to effectively manipulate and conditionally
compute summary statistics over subsets of a “big” dataset containing
many observations.</p>
<p><strong>Recommended reading:</strong> Review the <a
href="http://r4ds.had.co.nz/tibbles.html#introduction-4"><em>Introduction</em>
(10.1)</a> and <a
href="http://r4ds.had.co.nz/tibbles.html#tibbles-vs.data.frame"><em>Tibbles
vs. data.frame</em> (10.3)</a> sections of the <a
href="http://r4ds.had.co.nz/tibbles.html"><strong><em>R for Data
Science</em> book</strong></a>. We will initially be using the
<code>read_*</code> functions from the <a
href="http://readr.tidyverse.org/"><strong>readr</strong> package</a>.
These functions load data into a <em>tibble</em> instead of R’s
traditional data.frame. Tibbles are data frames, but they tweak some
older behaviors to make life a little easier. These sections explain the
few key small differences between traditional data.frames and
tibbles.</p>
<div id="review" class="section level2">
<h2>Review</h2>
<div id="the-data" class="section level3">
<h3>The data</h3>
<p>We are going to use a yeast gene expression dataset. This is a
cleaned up version of a gene expression dataset from <a
href="http://www.ncbi.nlm.nih.gov/pubmed/17959824">Brauer et
al. Coordination of Growth Rate, Cell Cycle, Stress Response, and
Metabolic Activity in Yeast (2008) <em>Mol Biol Cell</em>
19:352-367</a>. These data are from a gene expression microarray, and in
this paper the authors are examining the relationship between growth
rate and gene expression in yeast cultures limited by one of six
different nutrients (glucose, leucine, ammonium, sulfate, phosphate,
uracil). If you give yeast a rich media loaded with nutrients except
restrict the supply of a <em>single</em> nutrient, you can control the
growth rate to any rate you choose. By starving yeast of specific
nutrients you can find genes that:</p>
<ol style="list-style-type: decimal">
<li><strong>Raise or lower their expression in response to growth
rate</strong>. Growth-rate dependent expression patterns can tell us a
lot about cell cycle control, and how the cell responds to stress. The
authors found that expression of &gt;25% of all yeast genes is linearly
correlated with growth rate, independent of the limiting nutrient. They
also found that the subset of negatively growth-correlated genes is
enriched for peroxisomal functions, and positively correlated genes
mainly encode ribosomal functions.</li>
<li><strong>Respond differently when different nutrients are being
limited</strong>. If you see particular genes that respond very
differently when a nutrient is sharply restricted, these genes might be
involved in the transport or metabolism of that specific nutrient.</li>
</ol>
<p>The file is called <a
href="https://thriv.github.io/biodatasci2018/data/brauer2007_tidy.csv"><strong>brauer2007_tidy.csv</strong></a>.
Later on we’ll actually start with the original raw data (minimally
processed) and manipulate it so that we can make it more amenable for
analysis.</p>
</div>
<div id="reading-in-data" class="section level3">
<h3>Reading in data</h3>
<p>We need to load both the dplyr and readr packages for efficiently
reading in and displaying this data. We’re also going to use many other
functions from the dplyr package. Make sure you have these packages
installed as described on the <a href="setup.html">setup page</a>.</p>
<pre class="r"><code># Load packages
library(readr)
library(dplyr)

# Read in data
ydat &lt;- read.csv(&#39;https://raw.githubusercontent.com/bioconnector/workshops/master/data/brauer2007_tidy.csv&#39;)

# Display the data
ydat

# Optionally, bring up the data in a viewer window
# View(ydat)</code></pre>
<pre><code>##    symbol systematic_name nutrient rate expression
## 1    SFB2         YNL049C  Glucose 0.05      -0.24
## 2    &lt;NA&gt;         YNL095C  Glucose 0.05       0.28
## 3    QRI7         YDL104C  Glucose 0.05      -0.02
## 4    CFT2         YLR115W  Glucose 0.05      -0.33
## 5    SSO2         YMR183C  Glucose 0.05       0.05
## 6    PSP2         YML017W  Glucose 0.05      -0.69
## 7    RIB2         YOL066C  Glucose 0.05      -0.55
## 8   VMA13         YPR036W  Glucose 0.05      -0.75
## 9    EDC3         YEL015W  Glucose 0.05      -0.24
## 10   VPS5         YOR069W  Glucose 0.05      -0.16
## 11   &lt;NA&gt;         YOL029C  Glucose 0.05      -0.22
## 12   AMN1         YBR158W  Glucose 0.05       0.18
## 13  SCW11         YGL028C  Glucose 0.05      -0.67
## 14   DSE2         YHR143W  Glucose 0.05      -0.59
## 15  COX15         YER141W  Glucose 0.05      -0.28
## 16   SPE1         YKL184W  Glucose 0.05      -0.19
## 17   MTF1         YMR228W  Glucose 0.05      -0.42
## 18   KSS1         YGR040W  Glucose 0.05      -0.76
## 19   &lt;NA&gt;         YHR036W  Glucose 0.05      -0.91
## 20   &lt;NA&gt;         YNL158W  Glucose 0.05      -0.47
## 21   YAP7         YOL028C  Glucose 0.05      -0.51
## 22   &lt;NA&gt;         YBR074W  Glucose 0.05      -1.01
## 23   YVC1         YOR087W  Glucose 0.05      -0.40
## 24  CDC40         YDR364C  Glucose 0.05      -0.19
## 25   &lt;NA&gt;         YPL162C  Glucose 0.05      -0.10
## 26   RMD1         YDL001W  Glucose 0.05      -0.22
## 27   PCL6         YER059W  Glucose 0.05      -0.25
## 28    AI4           Q0065  Glucose 0.05      -0.36
##                                                                      bp
## 1                                                 ER to Golgi transport
## 2                                            biological process unknown
## 3                                          proteolysis and peptidolysis
## 4                                               mRNA polyadenylylation*
## 5                                                       vesicle fusion*
## 6                                            biological process unknown
## 7                                               riboflavin biosynthesis
## 8                                                vacuolar acidification
## 9                                 deadenylylation-independent decapping
## 10                                          protein retention in Golgi*
## 11                                           biological process unknown
## 12                            negative regulation of exit from mitosis*
## 13                                cytokinesis, completion of separation
## 14                               cell wall organization and biogenesis*
## 15                               cytochrome c oxidase complex assembly*
## 16                                           pantothenate biosynthesis*
## 17                            transcription from mitochondrial promoter
## 18                               invasive growth (sensu Saccharomyces)*
## 19                                           biological process unknown
## 20                                           biological process unknown
## 21 positive regulation of transcription from RNA polymerase II promoter
## 22                                         proteolysis and peptidolysis
## 23                                                   cation homeostasis
## 24                              nuclear mRNA splicing, via spliceosome*
## 25                                           biological process unknown
## 26                                           biological process unknown
## 27                                 regulation of glycogen biosynthesis*
## 28                                                        RNA splicing*
##                                                                            mf
## 1                                                  molecular function unknown
## 2                                                  molecular function unknown
## 3                                               metalloendopeptidase activity
## 4                                                                 RNA binding
## 5                                                            t-SNARE activity
## 6                                                  molecular function unknown
## 7                                          pseudouridylate synthase activity*
## 8                 hydrogen-transporting ATPase activity, rotational mechanism
## 9                                                  molecular function unknown
## 10                                               protein transporter activity
## 11                                                 molecular function unknown
## 12                                                            protein binding
## 13                                       glucan 1,3-beta-glucosidase activity
## 14                                       glucan 1,3-beta-glucosidase activity
## 15 oxidoreductase activity, acting on NADH or NADPH, heme protein as acceptor
## 16                                           ornithine decarboxylase activity
## 17                 S-adenosylmethionine-dependent methyltransferase activity*
## 18                                                        MAP kinase activity
## 19                                                 molecular function unknown
## 20                                                 molecular function unknown
## 21                            RNA polymerase II transcription factor activity
## 22                                              metalloendopeptidase activity
## 23                                                  calcium channel activity*
## 24               RNA splicing factor activity, transesterification mechanism*
## 25                                                 molecular function unknown
## 26                                                 molecular function unknown
## 27                         cyclin-dependent protein kinase regulator activity
## 28                                                      endonuclease activity
##  [ reached &#39;max&#39; / getOption(&quot;max.print&quot;) -- omitted 198402 rows ]</code></pre>
</div>
</div>
<div id="the-dplyr-package" class="section level2">
<h2>The dplyr package</h2>
<p>The <a href="https://github.com/hadley/dplyr">dplyr package</a> is a
relatively new R package that makes data manipulation fast and easy. It
imports functionality from another package called magrittr that allows
you to chain commands together into a pipeline that will completely
change the way you write R code such that you’re writing code the way
you’re thinking about the problem.</p>
<p>When you read in data with the readr package
(<code>read_csv()</code>) and you had the dplyr package loaded already,
the data frame takes on this “special” class of data frames called a
<code>tbl</code> (pronounced “tibble”), which you can see with
<code>class(ydat)</code>. If you have other “regular” data frames in
your workspace, the <code>as_tibble()</code> function will convert it
into the special dplyr <code>tbl</code> that displays nicely (e.g.:
<code>iris &lt;- as_tibble(iris)</code>). You don’t have to turn all
your data frame objects into tibbles, but it does make working with
large datasets a bit easier.</p>
<p>You can read more about tibbles in <a
href="http://r4ds.had.co.nz/tibbles.html">Tibbles chapter in <em>R for
Data Science</em></a> or in the <a
href="https://cran.r-project.org/web/packages/tibble/vignettes/tibble.html">tibbles
vignette</a>. They keep most of the features of data frames, and drop
the features that used to be convenient but are now frustrating
(i.e. converting character vectors to factors). You can read more about
the differences between data frames and tibbles in <a
href="https://cran.r-project.org/web/packages/tibble/vignettes/tibble.html#tibbles-vs-data-frames">this
section of the tibbles vignette</a>, but the major convenience for us
concerns <strong>printing</strong> (aka displaying) a tibble to the
screen. When you print (i.e., display) a tibble, it only shows the first
10 rows and all the columns that fit on one screen. It also prints an
abbreviated description of the column type. You can control the default
appearance with options:</p>
<ul>
<li><code>options(tibble.print_max = n, tibble.print_min = m)</code>: if
there are more than <em>n</em> rows, print only the first <em>m</em>
rows. Use <code>options(tibble.print_max = Inf)</code> to always show
all rows.</li>
<li><code>options(tibble.width = Inf)</code> will always print all
columns, regardless of the width of the screen.</li>
</ul>
</div>
<div id="dplyr-verbs" class="section level2">
<h2>dplyr verbs</h2>
<p>The dplyr package gives you a handful of useful
<strong>verbs</strong> for managing data. On their own they don’t do
anything that base R can’t do. Here are some of the
<em>single-table</em> verbs we’ll be working with in this lesson
(single-table meaning that they only work on a single table – contrast
that to <em>two-table</em> verbs used for joining data together, which
we’ll cover in a later lesson).</p>
<ol style="list-style-type: decimal">
<li><code>filter()</code></li>
<li><code>select()</code></li>
<li><code>mutate()</code></li>
<li><code>arrange()</code></li>
<li><code>summarize()</code></li>
<li><code>group_by()</code></li>
</ol>
<p>They all take a data frame or tibble as their input for the first
argument, and they all return a data frame or tibble as output.</p>
<div id="filter" class="section level3">
<h3>filter()</h3>
<p>If you want to filter <strong>rows</strong> of the data where some
condition is true, use the <code>filter()</code> function.</p>
<ol style="list-style-type: decimal">
<li>The first argument is the data frame you want to filter,
e.g. <code>filter(mydata, ...</code>.</li>
<li>The second argument is a condition you must satisfy,
e.g. <code>filter(ydat, symbol == "LEU1")</code>. If you want to satisfy
<em>all</em> of multiple conditions, you can use the “and” operator,
<code>&amp;</code>. The “or” operator <code>|</code> (the pipe
character, usually shift-backslash) will return a subset that meet
<em>any</em> of the conditions.</li>
</ol>
<ul>
<li><code>==</code>: Equal to</li>
<li><code>!=</code>: Not equal to</li>
<li><code>&gt;</code>, <code>&gt;=</code>: Greater than, greater than or
equal to</li>
<li><code>&lt;</code>, <code>&lt;=</code>: Less than, less than or equal
to</li>
</ul>
<p>Let’s try it out. For this to work you have to have already loaded
the dplyr package. Let’s take a look at <a
href="http://www.yeastgenome.org/locus/Leu1/overview">LEU1</a>, a gene
involved in leucine synthesis.</p>
<pre class="r"><code># First, make sure you&#39;ve loaded the dplyr package
library(dplyr)

# Look at a single gene involved in leucine synthesis pathway
filter(ydat, symbol == &quot;LEU1&quot;)

# Optionally, bring that result up in a View window
# View(filter(ydat, symbol == &quot;LEU1&quot;))

# Look at multiple genes
filter(ydat, symbol==&quot;LEU1&quot; | symbol==&quot;ADH2&quot;)

# Look at LEU1 expression at a low growth rate due to nutrient depletion
# Notice how LEU1 is highly upregulated when leucine is depleted!
filter(ydat, symbol==&quot;LEU1&quot; &amp; rate==.05)

# But expression goes back down when the growth/nutrient restriction is relaxed
filter(ydat, symbol==&quot;LEU1&quot; &amp; rate==.3)

# Show only stats for LEU1 and Leucine depletion. 
# LEU1 expression starts off high and drops
filter(ydat, symbol==&quot;LEU1&quot; &amp; nutrient==&quot;Leucine&quot;)

# What about LEU1 expression with other nutrients being depleted?
filter(ydat, symbol==&quot;LEU1&quot; &amp; nutrient==&quot;Glucose&quot;)</code></pre>
<p>Let’s look at this graphically. Don’t worry about what these commands
are doing just yet - we’ll cover that later on when we talk about
ggplot2. Here’s I’m taking the filtered dataset containing just
expression estimates for LEU1 where I have 36 rows (one for each of 6
nutrients <span class="math inline">\(\times\)</span> 6 growth rates),
and I’m <em>piping</em> that dataset to the plotting function, where I’m
plotting rate on the x-axis, expression on the y-axis, mapping the value
of nutrient to the color, and using a line plot to display the data.</p>
<pre class="r"><code>library(ggplot2)
filter(ydat, symbol==&quot;LEU1&quot;) %&gt;% 
  ggplot(aes(rate, expression, colour=nutrient)) + geom_line(lwd=1.5)</code></pre>
<p><img src="dplyr_dataframes_files/figure-html/demo_plot_expr_over_rate_singlegene-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Look closely at that! LEU1 is <em>highly expressed</em> when starved
of leucine because the cell has to synthesize its own! And as the amount
of leucine in the environment (the growth <em>rate</em>) increases, the
cell can worry less about synthesizing leucine, so LEU1 expression goes
back down. Consequently the cell can devote more energy into other
functions, and we see other genes’ expression very slightly raising.</p>
<hr />
<p><strong>EXERCISE 1</strong></p>
<ol style="list-style-type: decimal">
<li>Display the data where the gene ontology biological process (the
<code>bp</code> variable) is “leucine biosynthesis” (case-sensitive)
<em>and</em> the limiting nutrient was Leucine. (Answer should return a
24-by-7 data frame – 4 genes <span class="math inline">\(\times\)</span>
6 growth rates).</li>
<li>Gene/rate combinations had high expression (in the top 1% of
expressed genes)? <em>Hint:</em> see <code>?quantile</code> and try
<code>quantile(ydat$expression, probs=.99)</code> to see the expression
value which is higher than 99% of all the data, then
<code>filter()</code> based on that. Try wrapping your answer with a
<code>View()</code> function so you can see the whole thing. What does
it look like those genes are doing? Answer should return a 1971-by-7
data frame.</li>
</ol>
<hr />
<div id="aside-writing-data-to-file" class="section level4">
<h4>Aside: Writing Data to File</h4>
<p>What we’ve done up to this point is read in data from a file
(<code>read_csv(...)</code>), and assigning that to an object in our
<em>workspace</em> (<code>ydat &lt;- ...</code>). When we run operations
like <code>filter()</code> on our data, consider two things:</p>
<ol style="list-style-type: decimal">
<li>The <code>ydat</code> object in our workspace is not being modified
directly. That is, we can <code>filter(ydat, ...)</code>, and a result
is returned to the screen, but <code>ydat</code> remains the same. This
effect is similar to what we demonstrated in our first session.</li>
</ol>
<pre class="r"><code># Assign the value &#39;50&#39; to the weight object.
weight &lt;- 50

# Print out weight to the screen (50)
weight

# What&#39;s the value of weight plus 10?
weight + 10

# Weight is still 50
weight

# Weight is only modified if we *reassign* weight to the modified value
weight &lt;- weight+10
# Weight is now 60
weight</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>More importantly, the <em>data file on disk</em>
(<code>data/brauer2007_tidy.csv</code>) is <em>never</em> modified. No
matter what we do to ydat, the file is never modified. If we want to
<em>save</em> the result of an operation to a file on disk, we can
assign the result of an operation to an object, and
<code>write_csv</code> that object to disk. See the help for
<code>?write_csv</code> (note, <code>write_csv()</code> with an
underscore is part of the <strong>readr</strong> package – not to be
confused with the built-in <code>write.csv()</code> function).</li>
</ol>
<pre class="r"><code># What&#39;s the result of this filter operation?
filter(ydat, nutrient==&quot;Leucine&quot; &amp; bp==&quot;leucine biosynthesis&quot;)

# Assign the result to a new object
leudat &lt;- filter(ydat, nutrient==&quot;Leucine&quot; &amp; bp==&quot;leucine biosynthesis&quot;)

# Write that out to disk
write_csv(leudat, &quot;leucinedata.csv&quot;)</code></pre>
<p>Note that this is different than saving your <em>entire workspace to
an Rdata file</em>, which would contain all the objects we’ve created
(weight, ydat, leudat, etc).</p>
</div>
</div>
<div id="select" class="section level3">
<h3>select()</h3>
<p>The <code>filter()</code> function allows you to return only certain
<em>rows</em> matching a condition. The <code>select()</code> function
returns only certain <em>columns</em>. The first argument is the data,
and subsequent arguments are the columns you want.</p>
<pre class="r"><code># Select just the symbol and systematic_name
select(ydat, symbol, systematic_name)

# Alternatively, just remove columns. Remove the bp and mf columns.
select(ydat, -bp, -mf)

# Notice that the original data doesn&#39;t change!
ydat</code></pre>
<p>Notice above how the original data doesn’t change. We’re selecting
out only certain columns of interest and throwing away columns we don’t
care about. If we wanted to <em>keep</em> this data, we would need to
<em>reassign</em> the result of the <code>select()</code> operation to a
new object. Let’s make a new object called <code>nogo</code> that does
not contain the GO annotations. Notice again how the original data is
unchanged.</p>
<pre class="r"><code># create a new dataset without the go annotations.
nogo &lt;- select(ydat, -bp, -mf)
nogo

# we could filter this new dataset
filter(nogo, symbol==&quot;LEU1&quot; &amp; rate==.05)

# Notice how the original data is unchanged - still have all 7 columns
ydat</code></pre>
</div>
<div id="mutate" class="section level3">
<h3>mutate()</h3>
<p>The <code>mutate()</code> function adds new columns to the data.
Remember, it doesn’t actually modify the data frame you’re operating on,
and the result is transient unless you assign it to a new object or
reassign it back to itself (generally, not always a good practice).</p>
<p>The expression level reported here is the <span
class="math inline">\(log_2\)</span> of the sample signal divided by the
signal in the reference channel, where the reference RNA for all samples
was taken from the glucose-limited chemostat grown at a dilution rate of
0.25 <span class="math inline">\(h^{-1}\)</span>. Let’s mutate this data
to add a new variable called “signal” that’s the actual raw signal ratio
instead of the log-transformed signal.</p>
<pre class="r"><code>mutate(nogo, signal=2^expression)</code></pre>
<p>Mutate has a nice little feature too in that it’s “lazy.” You can
mutate and add one variable, then continue mutating to add more
variables based on that variable. Let’s make another column that’s the
square root of the signal ratio.</p>
<pre class="r"><code>mutate(nogo, signal=2^expression, sigsr=sqrt(signal))</code></pre>
<p>Again, don’t worry about the code here to make the plot – we’ll learn
about this later. Why do you think we log-transform the data prior to
analysis?</p>
<pre class="r"><code>library(tidyr)
mutate(nogo, signal=2^expression, sigsr=sqrt(signal)) %&gt;% 
  gather(unit, value, expression:sigsr) %&gt;% 
  ggplot(aes(value)) + geom_histogram(bins=100) + facet_wrap(~unit, scales=&quot;free&quot;)</code></pre>
<p><img src="dplyr_dataframes_files/figure-html/demo_plot_expression_units-1.png" width="960" /></p>
</div>
<div id="arrange" class="section level3">
<h3>arrange()</h3>
<p>The <code>arrange()</code> function does what it sounds like. It
takes a data frame or tbl and arranges (or sorts) by column(s) of
interest. The first argument is the data, and subsequent arguments are
columns to sort on. Use the <code>desc()</code> function to arrange by
descending.</p>
<pre class="r"><code># arrange by gene symbol
arrange(ydat, symbol)

# arrange by expression (default: increasing)
arrange(ydat, expression)

# arrange by decreasing expression
arrange(ydat, desc(expression))</code></pre>
<hr />
<p><strong>EXERCISE 2</strong></p>
<ol style="list-style-type: decimal">
<li>First, re-run the command you used above to filter the data for
genes involved in the “leucine biosynthesis” biological process
<em>and</em> where the limiting nutrient is Leucine.</li>
<li>Wrap this entire filtered result with a call to
<code>arrange()</code> where you’ll arrange the result of #1 by the gene
symbol.</li>
<li>Wrap this entire result in a <code>View()</code> statement so you
can see the entire result.</li>
</ol>
<hr />
</div>
<div id="summarize" class="section level3">
<h3>summarize()</h3>
<p>The <code>summarize()</code> function summarizes multiple values to a
single value. On its own the <code>summarize()</code> function doesn’t
seem to be all that useful. The dplyr package provides a few convenience
functions called <code>n()</code> and <code>n_distinct()</code> that
tell you the number of observations or the number of distinct values of
a particular variable.</p>
<p>Notice that summarize takes a data frame and returns a data frame. In
this case it’s a 1x1 data frame with a single row and a single column.
The name of the column, by default is whatever the expression was used
to summarize the data. This usually isn’t pretty, and if we wanted to
work with this resulting data frame later on, we’d want to name that
returned value something easier to deal with.</p>
<pre class="r"><code># Get the mean expression for all genes
summarize(ydat, mean(expression))

# Use a more friendly name, e.g., meanexp, or whatever you want to call it.
summarize(ydat, meanexp=mean(expression))

# Measure the correlation between rate and expression 
summarize(ydat, r=cor(rate, expression))

# Get the number of observations
summarize(ydat, n())

# The number of distinct gene symbols in the data 
summarize(ydat, n_distinct(symbol))</code></pre>
</div>
<div id="group_by" class="section level3">
<h3>group_by()</h3>
<p>We saw that <code>summarize()</code> isn’t that useful on its own.
Neither is <code>group_by()</code> All this does is takes an existing
data frame and coverts it into a grouped data frame where operations are
performed by group.</p>
<pre class="r"><code>ydat
group_by(ydat, nutrient)
group_by(ydat, nutrient, rate)</code></pre>
<p>The real power comes in where <code>group_by()</code> and
<code>summarize()</code> are used together. First, write the
<code>group_by()</code> statement. Then wrap the result of that with a
call to <code>summarize()</code>.</p>
<pre class="r"><code># Get the mean expression for each gene
# group_by(ydat, symbol)
summarize(group_by(ydat, symbol), meanexp=mean(expression))

# Get the correlation between rate and expression for each nutrient
# group_by(ydat, nutrient)
summarize(group_by(ydat, nutrient), r=cor(rate, expression))</code></pre>
</div>
</div>
<div id="the-pipe" class="section level2">
<h2>The pipe: <strong>%&gt;%</strong></h2>
<div id="how-works" class="section level3">
<h3>How %&gt;% works</h3>
<p>This is where things get awesome. The dplyr package imports
functionality from the <a
href="https://github.com/smbache/magrittr">magrittr</a> package that
lets you <em>pipe</em> the output of one function to the input of
another, so you can avoid nesting functions. It looks like this:
<strong><code>%&gt;%</code></strong>. You don’t have to load the
magrittr package to use it since dplyr imports its functionality when
you load the dplyr package.</p>
<p>Here’s the simplest way to use it. Remember the <code>tail()</code>
function. It expects a data frame as input, and the next argument is the
number of lines to print. These two commands are identical:</p>
<pre class="r"><code>tail(ydat, 5)
ydat %&gt;% tail(5)</code></pre>
<p>Let’s use one of the dplyr verbs.</p>
<pre class="r"><code>filter(ydat, nutrient==&quot;Leucine&quot;)
ydat %&gt;% filter(nutrient==&quot;Leucine&quot;)</code></pre>
</div>
<div id="nesting-versus" class="section level3">
<h3>Nesting versus %&gt;%</h3>
<p>So what?</p>
<p>Now, think about this for a minute. What if we wanted to get the
correlation between the growth rate and expression separately for each
limiting nutrient only for genes in the leucine biosynthesis pathway,
and return a sorted list of those correlation coeffients rounded to two
digits? Mentally we would do something like this:</p>
<ol start="0" style="list-style-type: decimal">
<li>Take the <code>ydat</code> dataset</li>
<li><em>then</em> <code>filter()</code> it for genes in the leucine
biosynthesis pathway</li>
<li><em>then</em> <code>group_by()</code> the limiting nutrient</li>
<li><em>then</em> <code>summarize()</code> to get the correlation
(<code>cor()</code>) between rate and expression</li>
<li><em>then</em> <code>mutate()</code> to round the result of the above
calculation to two significant digits</li>
<li><em>then</em> <code>arrange()</code> by the rounded correlation
coefficient above</li>
</ol>
<p>But in code, it gets ugly. First, take the <code>ydat</code>
dataset</p>
<pre class="r"><code>ydat</code></pre>
<p><em>then</em> <code>filter()</code> it for genes in the leucine
biosynthesis pathway</p>
<pre class="r"><code>filter(ydat, bp==&quot;leucine biosynthesis&quot;)</code></pre>
<p><em>then</em> <code>group_by()</code> the limiting nutrient</p>
<pre class="r"><code>group_by(filter(ydat, bp==&quot;leucine biosynthesis&quot;), nutrient)</code></pre>
<p><em>then</em> <code>summarize()</code> to get the correlation
(<code>cor()</code>) between rate and expression</p>
<pre class="r"><code>summarize(group_by(filter(ydat, bp == &quot;leucine biosynthesis&quot;), nutrient), r = cor(rate,
    expression))</code></pre>
<p><em>then</em> <code>mutate()</code> to round the result of the above
calculation to two significant digits</p>
<pre class="r"><code>mutate(summarize(group_by(filter(ydat, bp == &quot;leucine biosynthesis&quot;), nutrient),
    r = cor(rate, expression)), r = round(r, 2))</code></pre>
<p><em>then</em> <code>arrange()</code> by the rounded correlation
coefficient above</p>
<pre class="r"><code>arrange(
  mutate(
    summarize(
      group_by(
        filter(ydat, bp==&quot;leucine biosynthesis&quot;), 
      nutrient), 
    r=cor(rate, expression)), 
  r=round(r, 2)), 
r)</code></pre>
<p>Now compare that with the mental process of what you’re actually
trying to accomplish. The way you would do this without pipes is
completely inside-out and backwards from the way you express in words
and in thought what you want to do. The pipe operator
<code>%&gt;%</code> allows you to pass the output data frame from one
function to the input data frame to another function.</p>
<p>This is how we would do that in code. It’s as simple as replacing the
word “then” in words to the symbol <code>%&gt;%</code> in code. (There’s
a keyboard shortcut that I’ll use frequently to insert the
<code>%&gt;%</code> sequence – you can see what it is by clicking the
<em>Tools</em> menu in RStudio, then selecting <em>Keyboard Shortcut
Help</em>. On Mac, it’s CMD-SHIFT-M.)</p>
<pre class="r"><code>ydat %&gt;% 
  filter(bp==&quot;leucine biosynthesis&quot;) %&gt;%
  group_by(nutrient) %&gt;% 
  summarize(r=cor(rate, expression)) %&gt;% 
  mutate(r=round(r,2)) %&gt;% 
  arrange(r)</code></pre>
</div>
<div id="piping-exercises" class="section level3">
<h3>Piping exercises</h3>
<p><strong>EXERCISE 3</strong></p>
<p>Here’s a warm-up round. Try the following.</p>
<p>Show the limiting nutrient and expression values for the gene ADH2
when the growth rate is restricted to 0.05. <em>Hint:</em> 2 pipes:
<code>filter</code> and <code>select</code>.</p>
<pre class="r"><code>ydat %&gt;% filter(symbol==&quot;ADH2&quot; &amp; rate==0.05) %&gt;% select(nutrient, expression)</code></pre>
<p>What are the four most highly expressed genes when the growth rate is
restricted to 0.05 by restricting glucose? Show only the symbol,
expression value, and GO terms. <em>Hint:</em> 4 pipes:
<code>filter</code>, <code>arrange</code>, <code>head</code>, and
<code>select</code>.</p>
<pre class="r"><code>ydat %&gt;% 
  filter(nutrient==&quot;Glucose&quot; &amp; rate==.05) %&gt;% 
  arrange(desc(expression)) %&gt;% 
  head(4) %&gt;% 
  select(symbol, expression, bp, mf) </code></pre>
<p>When the growth rate is restricted to 0.05, what is the average
expression level across all genes in the “response to stress” biological
process, separately for each limiting nutrient? What about genes in the
“protein biosynthesis” biological process? <em>Hint:</em> 3 pipes:
<code>filter</code>, <code>group_by</code>, <code>summarize</code>.</p>
<pre class="r"><code>ydat %&gt;% 
  filter(rate==.05 &amp; bp==&quot;response to stress&quot;) %&gt;% 
  group_by(nutrient) %&gt;% 
  summarize(meanexp=mean(expression))

ydat %&gt;% 
  filter(rate==.05 &amp; bp==&quot;protein biosynthesis&quot;) %&gt;% 
  group_by(nutrient) %&gt;% 
  summarize(meanexp=mean(expression))</code></pre>
<hr />
<p><strong>EXERCISE 4</strong></p>
<p>That was easy, right? How about some tougher ones.</p>
<p>First, some review. How do we see the number of distinct values of a
variable? Use <code>n_distinct()</code> within a
<code>summarize()</code> call.</p>
<pre class="r"><code>ydat %&gt;% summarize(n_distinct(mf))</code></pre>
<p>Which 10 biological process annotations have the most genes
associated with them? What about molecular functions? <em>Hint:</em> 4
pipes: <code>group_by</code>, <code>summarize</code> with
<code>n_distinct</code>, <code>arrange</code>, <code>head</code>.</p>
<pre class="r"><code>ydat %&gt;% 
  group_by(bp) %&gt;% 
  summarize(n=n_distinct(symbol)) %&gt;% 
  arrange(desc(n)) %&gt;% 
  head(10)

ydat %&gt;% 
  group_by(mf) %&gt;% 
  summarize(n=n_distinct(symbol)) %&gt;% 
  arrange(desc(n)) %&gt;% 
  head(10)</code></pre>
<p>How many distinct genes are there where we know what process the gene
is involved in but we don’t know what it does? <em>Hint:</em> 3 pipes;
<code>filter</code> where
<code>bp!="biological process unknown" &amp; mf=="molecular function unknown"</code>,
and after <code>select</code>ing columns of interest, pipe the output to
<code>distinct()</code>. The answer should be <strong>737</strong>, and
here are a few:</p>
<pre class="r"><code>ydat %&gt;% 
  filter(bp!=&quot;biological process unknown&quot; &amp; mf==&quot;molecular function unknown&quot;) %&gt;% 
  select(symbol, bp, mf) %&gt;% 
  distinct()</code></pre>
<p>When the growth rate is restricted to 0.05 by limiting Glucose, which
biological processes are the most upregulated? Show a sorted list with
the most upregulated BPs on top, displaying the biological process and
the average expression of all genes in that process rounded to two
digits. <em>Hint:</em> 5 pipes: <code>filter</code>,
<code>group_by</code>, <code>summarize</code>, <code>mutate</code>,
<code>arrange</code>.</p>
<pre class="r"><code>ydat %&gt;% 
  filter(nutrient==&quot;Glucose&quot; &amp; rate==.05) %&gt;% 
  group_by(bp) %&gt;% 
  summarize(meanexp=mean(expression)) %&gt;% 
  mutate(meanexp=round(meanexp, 2)) %&gt;% 
  arrange(desc(meanexp))</code></pre>
<p>Group the data by limiting nutrient (primarily) then by biological
process. Get the average expression for all genes annotated with each
process, separately for each limiting nutrient, where the growth rate is
restricted to 0.05. Arrange the result to show the most upregulated
processes on top. The initial result will look like the result below.
Pipe this output to a <code>View()</code> statement. What’s going on?
Why didn’t the <code>arrange()</code> work? <em>Hint:</em> 5 pipes:
<code>filter</code>, <code>group_by</code>, <code>summarize</code>,
<code>arrange</code>, <code>View</code>.</p>
<pre class="r"><code>ydat %&gt;% 
  filter(rate==0.05) %&gt;% 
  group_by(nutrient, bp) %&gt;% 
  summarize(meanexp=mean(expression)) %&gt;% 
  arrange(desc(meanexp))</code></pre>
<p>Let’s try to further process that result to get only the top three
most upregulated biolgocal processes for each limiting nutrient. Google
search “dplyr first result within group.” You’ll need a
<code>filter(row_number()......)</code> in there somewhere.
<em>Hint:</em> 5 pipes: <code>filter</code>, <code>group_by</code>,
<code>summarize</code>, <code>arrange</code>,
<code>filter(row_number()...</code>. <em>Note:</em> dplyr’s pipe syntax
used to be <code>%.%</code> before it changed to <code>%&gt;%</code>. So
when looking around, you might still see some people use the old syntax.
Now if you try to use the old syntax, you’ll get a deprecation
warning.</p>
<pre class="r"><code>ydat %&gt;% 
  filter(rate==0.05) %&gt;% 
  group_by(nutrient, bp) %&gt;% 
  summarize(meanexp=mean(expression)) %&gt;% 
  arrange(desc(meanexp)) %&gt;% 
  filter(row_number()&lt;=3)</code></pre>
</div>
</div>

    <script>
        (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
        (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
        m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
        })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
        ga('create', 'UA-93043521-1', 'auto');
        ga('send', 'pageview');

         
        var links = document.querySelectorAll('a');
        Array.prototype.map.call(links, function(item) {
            if (item.host != document.location.host) {
                item.addEventListener('click', function() {
                    var action = item.getAttribute('data-action') || 'follow';
                    ga('send', 'event', 'outbound', action, item.href);
                });
            }
        });
    </script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
